### PR TITLE
Evaluation subpackage

### LINKED ISSUE #76: Scoring function for intervals of size one
The known anomalies in a time series could be of length one. In that case, the current segment based scoring function will not consider the existence of an overlap.

```
def _partition(expected, observed, start=None, end=None):
    edges = set()

    if start is not None:
        edges.add(start)

    if end is not None:
        edges.add(end)

    for edge in expected + observed:
        edges.update(edge)

    partitions = list()
    edges = sorted(edges)
    last = edges[0]
    for edge in edges[1:]:
        partitions.append((last, edge))
        last = edge

    expected_parts = list()
    observed_parts = list()
    weights = list()
    for part in partitions:
        weights.append(part[1] - part[0] + 1)
        expected_parts.append(_any_overlap(part, expected))
        observed_parts.append(_any_overlap(part, observed))

    return expected_parts, observed_parts, weights
```

If we look at the `_partition` function, we notice that `edges` is a set and will only include unique edges.
But an example of a perfectly detected anomalies can be as follows:
```
known_anomalies = 
| start | end |
|-------|-----|
| 3     | 3   |

observed anomalies = 
| start | end |
|-------|-----|
| 3     | 3   |
```

Now `edges = {3}`, and therefore, the result of this function will be an empty partition.

### LINKED ISSUE #83: Rename evaluation to benchmark and metrics to evaluation
Currently, the orion package is organized as follows. 
```
orion
├── evaluation.py 
└── metrics.py
```
With the new metrics subpackage, we would like to rename it to evaluation and update `evaluation.py` to `benchmark.py`.
```
orion
├── benchmark.py 
└── orion/evaluation -> used to be metrics.py
    ├── __init__.py 
    ├── common.py 
    ├── contextual.py 
    ├── point.py 
    └── utils.py
```

### PR DESCRIPTION
**Pull request by sarahmish**
Implement the Evaluation subpackage
Resolve #76 : addressing point anomalies in evaluation
Resolve #83 : renaming modules.

### CODE REVIEW

File: orion/eval/evaluation.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
+import numpy as np
+from sklearn import metrics
+
+
+class Evaluator:
```
Comment (csala) [inline orion/eval/evaluation.py:5]: I don't think a class is needed here, since there is nothing from the instance itself that needs to be used other than functions.

I would remove the classes and simply make them functions (which use other functions).

More specifically, I would organize this subpackage as follows:

```
orion/evaluation
├── __init__.py  -> Import the end-user functions from contextual and point
├── common.py -> Shared functions used in the contextual and point modules, including _accuracy, ...
├── contextual.py -> Import common functions and define get_contextual_confusion_matrix and contextual_accuracy, ...
├── point.py -> Import common functions and define get_point_confusion_matrix and point_accuracy...
└── utils.py
```

File: orion/eval/evaluation.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        for expected_seq in expected:
+            found = False
+            for observed_seq in observed:
+                if self._overlap(expected_seq, observed_seq):
+                    if not found:
+                        tp += 1
+                        found = True
+                    if observed_seq in observed_copy:
+                        observed_copy.remove(observed_seq)
+            if not found:
```
Comment (csala) [inline orion/eval/evaluation.py:81]: Can we add blank lines before reducing indentation?
This creates a clearer visual separation between blocks.
For example, here I would add them under lines 80 and 82, and optionally under 78 as well.

File: orion/eval/evaluation.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        fp += len(observed_copy)
+
+        return None, fp, fn, tp
+
+    def get_confusion_matrix(self, expected, observed, method=None,
+                             data=None, start=None, end=None):
+        """Compute the confusion matrix between the ground truth and the detected anomalies.
+
+        Args:
+            * expected (pd.DataFrame): Ground truth.
```
Comment (csala) [inline orion/eval/evaluation.py:92]: A couple of things should be considered regarding the docstrings:
- The `*` can be removed when listing the `Args`, so line starts directly with the argument name.
- A `Returns` block should be added listing the return type and a description in the next line.
- A blank line should always exist between the first line and the rest, and also before Args and Returns
- (Optional): I particularly like to put the argument description in a line below the argument name instead of putting all of it in the same line. This ensures that format is properly parsed when documentation is rendered on Sphinx if the description has more than one line.

```
"""Short description in a single line ending with a dot.

Longer description that can span across multiple lines. Longer
description that can span across multiple lines. Longer description
that can span across multiple lines.

Args:
    arg_name (arg_type):
        argument description.
    arg_name (arg_type):
        Argument description that spans across multiple lines. Argument
        description that spans across multiple lines. Argument description
        that spans across multiple lines.

Returns:
    return_type:
        description of the returned objects
```

**Comment by sarahmish** [comment]
Resolve #83.

File: orion/evaluation/contextual.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
+""" The evaluator module provides metrics to assess the perfomance of
+anomaly detection methods using classification metrics: accuracy, precision,
+recall, and f1 score. This class addresses the problem as contextual anomalies.
+
+anomaly = [(start, end)] inclusive interval.
+"""
+
+from .common import _accuracy, _any_overlap, _f1_score, _precision, _recall, _weighted_segment
```
Comment (csala) [inline orion/evaluation/contextual.py:8]: We should always use explicit absolute imports:

```python
from orion.evaluation.common import ...
```

This helps readability by indicating where to find the code of the imports at any moment without forcing to look at which the current module is.

File: orion/evaluation/contextual.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
+""" The evaluator module provides metrics to assess the perfomance of
```
Comment (csala) [inline orion/evaluation/contextual.py:1]: I think that this docstring is outdated.

File: orion/evaluation/point.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
+""" The evaluator module provides a mechanic to assess the perfomance of
```
Comment (csala) [inline orion/evaluation/point.py:1]: I think that this docstring is outdated.

File: orion/evaluation/utils.py [hunk 7]
```diff
@@ -5,12 +5,16 @@
 
 
 def from_pandas_contextual(df):
-    """ convert contextual `pd.DataFrame` to list of tuples.
+    """ Convert contextual ``pandas.DataFrame`` to list of tuples.
 
     Args:
-        * df (pd.DataFrame): contains start and end.
-    """
+        df (DataFrame): anomalies, passed as ``pandas.DataFrame``
```
Comment (csala) [inline orion/evaluation/utils.py:11]: The description here should be put in a different line:

```
df (DataFrame):
    anomalies, passed as ``pandas.DataFrame``
    containing two columns: start and stop.
```

When there is only one line in the description it does not really matter, but when you break the line in two or more, if the first one is kept at the same level as the arg name it does something funky when it is rendered by Sphinx.

File: orion/evaluation/utils.py [hunk 8]
```diff
@@ -5,12 +5,16 @@
...
+        containing two columns: start and stop.
 
+    Returns:
+        list:
+            tuple (start, end) timestamp.
```
Comment (csala) [inline orion/evaluation/utils.py:16]: I would make this description a bit more verbose, otherwise it can be confusing.
Something like:

```
list:
    list of tuples containing start and end timestamps.
```

Comment (csala) [reply orion/evaluation/utils.py:16]: Also, since the method has an explicit `rise` within it, it would be great to add a `Rises` entry indicating it.

File: tests/evaluation/test_common.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+def expected():
+    return (1, 5)
+
+
+@pytest.fixture
+def observed():
+    return (2, 5)
+
+
+def test__overlap(expected, observed):
```
Comment (csala) [inline tests/evaluation/test_common.py:17]: Are these tests enough?

Even though the test function is quite simple, I would I feel that here we would want a few test cases, especially the edge ones:
- no overlap
- no overlap, edges touch
- exact match
- one contained within the other, one edge matches
- one fully contained within the other, no edge match
- start == end (if this is something that could happen - current code seems to fail at this one)

Comment (csala) [reply tests/evaluation/test_common.py:17]: The same thing applies for the other functions being tested in this module.

File: orion/evaluation/contextual.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+            Function for computing confusion matrix (weighted segment, overlap segment).
+
+    Returns:
+        float:
+            F1 score between the ground truth and detected anomalies.
+    """
+    return _f1_score(expected, observed, data, start, end, method, contextual_confusion_matrix)
+
+
+METRICS = {
```
Comment (csala) [inline orion/evaluation/contextual.py:201]: I would rename this to `CONTEXTUAL_METRICS`, to be consistent with the point ones.

Also, maybe these two dictionaries could be directly defined in the parent `__init__`?

File: tests/evaluation/test_common.py (hunk 9 above)
Comment (sarahmish) [reply tests/evaluation/test_common.py]: Updated the unittests to include edge cases and edge cases with points

File: orion/evaluation/contextual.py (hunk 10 above)
Comment (sarahmish) [reply orion/evaluation/contextual.py]: Updated.

File: orion/evaluation/utils.py (hunk 8 above)
Comment (sarahmish) [reply orion/evaluation/utils.py:17]: Updated.

File: orion/evaluation/utils.py (hunk 7 above)
Comment (sarahmish) [reply orion/evaluation/utils.py]: Done, didn't know multilines were an issue

File: orion/evaluation/point.py (hunk 6 above)
Comment (sarahmish) [reply orion/evaluation/point.py]: Removed.

File: orion/evaluation/contextual.py (hunk 5 above)
Comment (sarahmish) [reply orion/evaluation/contextual.py]: Removed.

File: orion/evaluation/contextual.py (hunk 4 above)
Comment (sarahmish) [reply orion/evaluation/contextual.py]: Updated.

File: orion/evaluation/common.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+def _accuracy(expected, observed, data, start, end, method, get_confusion_matrix):
+    tn, fp, fn, tp = get_confusion_matrix(expected,
+                                          observed,
+                                          data,
+                                          start,
+                                          end,
+                                          method)
+    if tn is None:
+        raise ValueError("Cannot obtain accuracy score for overlap segment method.")
+    return (tp + tn) / (tn + fp + fn + tp)
```
[the remark this answers is not in the document]
Comment (sarahmish) [reply orion/evaluation/common.py:33]: Done.

File: orion/evaluation/common.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        if not found:
+            fn += 1
+
+    fp += len(observed_copy)
+
+    return None, fp, fn, tp
+
+
+def _accuracy(expected, observed, data, start, end, method, get_confusion_matrix):
+    tn, fp, fn, tp = get_confusion_matrix(expected,
```
[the remark this answers is not in the document]
Comment (sarahmish) [reply orion/evaluation/common.py]: Updated.

File: orion/evaluation/contextual.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+            number of true negative, false positive, false negative, true positive.
+    """
+
+    def ws(x, y, z, w):
+        return _weighted_segment(x, y, _contextual_partition, z, w)
+
+    if method is not None:
+        function = method
+    else:
+        function = ws
```
Comment (sarahmish) [inline orion/evaluation/contextual.py:93]: Is there a better approach to define this?

I basically want `_weighted_segment` to be the default and `_overlap_segment` to be chosen when the user specifies they want the other approach.

Comment (csala) [reply orion/evaluation/contextual.py:93]: Instead of using a `method` argument, I would just have a boolean argument called `weighted=True`, documented as "If True (default), consider overlap weights when computing metrics. Otherwise, use simple boolean overlapping indicators." or similar.
